feature/text_based.py

```diff
@@ -0,0 +1,694 @@
+            Optimization method used to perform content selection.
+            See TextBased in Selector.py for supported options definition
+
+        cost_metric: str, default = "diverse"
+
+        trials: int, default = 10
+
+        Returns
+        -------
+        List of indices in data that are selected.
+        """
+
+        # Process data
+        self.cost_metric = cost_metric
+        self.featurization_method = featurization_method
+        self.selection_size = selection_size
+        self._process_df(input_df, categories, selection_size, optimization_method, cost_metric)
+
+        # Run multi-level set covering optimization
+        if optimization_method == "exact":
+            selected = self._select_multi_level_optimization()
+        # Run KMeans selection algorithm
+        elif optimization_method == "kmeans":
+            selected = self._select_kmeans()
+        # Run greedy heuristic
+        elif optimization_method == "greedy":
+            selected = self._select_greedy()
+        # Run random selection algorithm
+        elif optimization_method == "random":
+            selected = self._select_random(trials=trials)
+        else:
+            raise NotImplementedError(f"{optimization_method} has not been implemented.")
+
+        # Sort selection
+        selected.sort()
+
+        return selected
+
+    def _process_df(self, input_df: pd.DataFrame, categories: List[int], selection_size: int,
+                    optimization_method: str, cost_metric: str) -> NoReturn:
+        """_process_df
+        We have implemented two versions of this function. In the version here, the categories parameter
+        actually refers to the labels (rather than categories) to be covered. It generates features for input data
+        using a specified featurization method, and creates a matrix where each row corresponds to a label and each
+        column corresponds to a feature.
+        The version of code for generating labels from categories has been moved to utils.py.
+
+        Parameters
+        ----------
+        input_df: pd.DataFrame
+            Input data frame with categories and features of content to select from.
+
+        categories: List[int]
+            List of columns in data that contains categories/labels to be covered.
+
+        selection_size: int
+            Number of feature to select.
+
+        optimization_method: str
+            The function need method to prevent performing featurization in
+             the case that optimization method is random(t or unicost) or greedy(unicost)
+
+        cost_metric: str
+            The function need cost metric to prevent performing featurization in
+             the case that cost metric is None or unicost (for random and greedy)
+
+        Returns
+        -------
+        No return
+       --------
+        """
+        self.matrix = categories.to_numpy()
+        self._num_rows, self._num_cols = self.matrix.shape
+        option_map = {
```

Comment:
what is this options map? Can we add some comments to this blocks of code plz

Reply:
Added some comments for options map

Reply:
OK. I spent quite a bit of time to understand this.. becase there is a lot happening.. there maps, lambdas, factory methods etc. 

IF I am correct, all we ware trying to do here is to decide whether we need text featurization or not. YES? 

IF so, you can very much simplift this (e.g., you don't need lambdas that are none jsut to pass and a bit of more explicit naming) 

```
from collections import defaultdict

is_featurization_needed = defaultdict(True) 
is_featurization_needed[("random", True, "diverse")] = False
is_featurization_needed[("random", False "unicost")] = False
is_featurization_needed[("random", True, "unicost")] = False
is_featurization_needed[("greedy", False "unicost")] = False
is_featurization_needed[("greedy", True, "unicost")] = False
is_featurization_needed[("exact", True, "unicost")] = False
is_featurization_needed[("exact", False "unicost")] = False

if is_featurization_needed: 
   feature_column = self.featurization_method.fit_transform(input_df)
````

Reply:
why is not (random, False, diverse) is not added as False?

Reply:
I am also thinking about converting all True -> False and False -> True in the is_featurization_needed dictionary 
so that the configuration can be as (opt_method, selection is None, cost_metric) to avoid the negation in the middle. 

Sth to consider, if you believe it makes it easier to parse/reason about?

Reply:
Higher level comment about this process_data function. I don't think it is needed. Back then, when dealing with categories, some processing was involved. Here the input "labels" are already in the format we want. 

So this part can totally be moved into the constuctor and removed from this function:
```
        self.matrix = labels.to_numpy()
        self._num_rows, self._num_cols = self.matrix.shape
```

THEN.. what remains here is text featurization. So I would turn this function around in the constructor and make it 

```
self.text_embeddings = _get_text_embeddings(selection_size, optimization_method, cost_metric) 
```

(notice I renamed features to text_embeddings.

Reply:
I agree with overall idea, expect these two points:

1- Move below part to constructor:
self.matrix = labels.to_numpy()
self._num_rows, self._num_cols = self.matrix.shape

If we move the code for setting self.matrix and self._num_rows and self._num_cols into the constructor, then we need to make sure that these values are set before calling _get_text_embeddings in the constructor. As a result, the inputs to the _TextBased class constructor will need to change to include labels, and any code that creates an instance of the _TextBased class will also need to include labels.
 Also, if we move the process of generating text embeddings to the constructor, we would need to add input_df as an input argument to the SelectionMethod.TextBased() constructor. This would also require adding input_df and labels as inputs to the ContentSelector class. This change would break the consistency of input arguments across different selective methods.

Therefore, moving those lines to constructor require changing the input arguments of the SelectionMethod.TextBased() constructor and adding input_df and labels as inputs to the ContentSelector class. It would also make storing labels and the number of rows and columns in the constructor unnecessary.

2- (random, False, diverse) -> rand(t=diverse()) when t=None. It mean solving a set cover with diverse cost which it requires featurization.

Reply:
#2 good there, thanks for clarifying. 

#1, I made a mistake. I didn't mean the constructor but lifting those lines from process() to where process() is called, which is run() (i thought we were in the init().. but I meant run()

So like this;
```
def run()
   ... 
    self.matrix = labels.to_numpy()
    self._num_rows, self._num_cols = self.matrix.shape
    self.text_embeddings = _get_text_embeddings(selection_size, optimization_method, cost_metric)
    ... 
````
This is the "same" as before except that process() can be renamed to get_text_embeddings() now and only does that. In fact there is no real processing anymore (unlike the category stuff before) we are just reading the matrix from labels

This simplifies/clarifies

Reply:
Make sense. Thank you for clarification! Done!

Reply:
Cool. I assume you will push an updated version soon

Reply:
Yes, I will do soon